File: 11-Recommender System/ex11.m
fprintf('Loading movie ratings dataset.\n\n');
load ('ex8_movies.mat');

fprintf('Average rating for movie 1 (Toy Story): %f / 5\n\n', mean(Y(1, R(1, :))));
imagesc(Y);
ylabel('Movies');
xlabel('Users');

fprintf('\nProgram paused. Press enter to continue.\n');
pause;

function [J, grad] = cofiCostFunc(params, Y, R, num_users, num_movies, num_features, lambda)
  X = reshape(params(1:num_movies*num_features), num_movies, num_features);
  Theta = reshape(params(num_movies*num_features+1:end), num_users, num_features);
  J = 0;
  X_grad = zeros(size(X));
  Theta_grad = zeros(size(Theta));
  D = (R .* (X * Theta' - Y));
  J = 1 / 2 * sum(sum( D .^ 2 ));
  J = J + (lambda / 2) * (sum(sumsq(Theta)) + sum(sumsq(X))); 
  for i = 1:num_movies
	  idx = find(R(i,:) == 1);
	  theta_tmp = Theta(idx,:);
	  y_tmp = Y(i,idx);
	  X_grad(i,:) = ((X(i,:) * theta_tmp') - y_tmp) * theta_tmp + lambda * X(i,:);
  endfor
  for j = 1:num_users
    idx = find(R(:,j) == 1);
	  x_tmp = X(idx,:);
	  y_tmp = Y(idx,j);
	  Theta_grad(j,:) = ((Theta(j,:) * x_tmp') - y_tmp') * x_tmp + lambda * Theta(j,:);
  endfor
  grad = [X_grad(:); Theta_grad(:)];
endfunction

function numgrad = computeNumericalGradient(J, theta)
  numgrad = zeros(3,38);
  perturb = zeros(size(theta));
  e = 1e-4;
  for p = 1:numel(theta)
    perturb(p) = e;
    loss1 = J(theta - perturb);
    loss2 = J(theta + perturb);
    numgrad(p) = (loss2 - loss1)/(2*e);
    perturb(p) = 0;
  endfor
endfunction

function checkCostFunction(lambda)
  if ~exist('lambda', 'var') || isempty(lambda)
    lambda = 0;
  endif
  X_t = rand(4, 3);
  Theta_t = rand(5, 3);
  Y = X_t * Theta_t';
  Y(rand(size(Y)) > 0.5) = 0;
  R = zeros(size(Y));
  R(Y ~= 0) = 1;
  X = randn(size(X_t));
  Theta = randn(size(Theta_t));
  num_users = size(Y, 2);
  num_movies = size(Y, 1);
  num_features = size(Theta_t, 2);
  numgrad = computeNumericalGradient(@(t) cofiCostFunc(t, Y, R, num_users, num_movies, num_features, lambda), [X(:); Theta(:)])
  size(numgrad)
  [cost, grad] = cofiCostFunc([X(:); Theta(:)],  Y, R, num_users, num_movies, num_features, lambda);
  size(grad)
  disp([numgrad grad]);
  fprintf(['The above two columns you get should be very similar. Left-Your Numerical Gradient, Right-Analytical Gradient)\n\n']);
  diff = norm(numgrad-grad)/norm(numgrad+grad);
endfunction

function [Ynorm, Ymean] = normalizeRatings(Y, R)
  [m, n] = size(Y);
  Ymean = zeros(m, 1);
  Ynorm = zeros(size(Y));
  for i = 1:m
    idx = find(R(i, :) == 1);
    Ymean(i) = mean(Y(i, idx));
    Ynorm(i, idx) = Y(i, idx) - Ymean(i);
  endfor
endfunction

function movieList = loadMovieList()
  fid = fopen('movie_ids.txt');
  n = 1682;  % Total number of movies 
  movieList = cell(n, 1);
  for i = 1:n
    line = fgets(fid);
    [idx, movieName] = strtok(line, ' ');
    movieList{i} = strtrim(movieName);
  endfor
  fclose(fid);
endfunction

load ('ex8_movieParams.mat');
%  Reduce the data set size so that this runs faster
num_users = 4; num_movies = 5; num_features = 3;
X = X(1:num_movies, 1:num_features);
Theta = Theta(1:num_users, 1:num_features);
Y = Y(1:num_movies, 1:num_users);
R = R(1:num_movies, 1:num_users);

J = cofiCostFunc([X(:) ; Theta(:)], Y, R, num_users, num_movies, num_features, 0);
fprintf('\nProgram paused. Press enter to continue.\n');
pause;
fprintf('\nChecking Gradients (without regularization) ... \n');
#checkCostFunction;

fprintf('\nProgram paused. Press enter to continue.\n');
pause;

J = cofiCostFunc([X(:) ; Theta(:)], Y, R, num_users, num_movies, num_features, 1.5);             
fprintf(['Cost at loaded parameters (lambda = 1.5): %f '], J);

fprintf('\nProgram paused. Press enter to continue.\n');
pause;
fprintf('\nChecking Gradients (with regularization) ... \n');
#checkCostFunction(1.5);

fprintf('\nProgram paused. Press enter to continue.\n');
pause;

movieList = loadMovieList();
my_ratings = zeros(1682, 1);
my_ratings(1) = 4;
my_ratings(98) = 2;
my_ratings(7) = 3;
my_ratings(12)= 5;
my_ratings(54) = 4;
my_ratings(64)= 5;
my_ratings(66)= 3;
my_ratings(69) = 5;
my_ratings(183) = 4;
my_ratings(226) = 5;
my_ratings(355)= 5;

fprintf('\n\nNew user ratings:\n');
for i = 1:length(my_ratings)
    if my_ratings(i) > 0 
        fprintf('Rated %d for %s\n', my_ratings(i), ...
                 movieList{i});
    endif
endfor

fprintf('\nProgram paused. Press enter to continue.\n');
pause;

fprintf('\nTraining collaborative filtering...\n');
load('ex8_movies.mat');
Y = [my_ratings Y];
R = [(my_ratings ~= 0) R];
[Ynorm, Ymean] = normalizeRatings(Y, R);
num_users = size(Y, 2);
num_movies = size(Y, 1);
num_features = 10;
X = randn(num_movies, num_features);
Theta = randn(num_users, num_features);
initial_parameters = [X(:); Theta(:)];
options = optimset('GradObj', 'on', 'MaxIter', 100);
lambda = 10;
theta = fmincg (@(t)(cofiCostFunc(t, Y, R, num_users, num_movies, num_features, lambda)), initial_parameters, options);
X = reshape(theta(1:num_movies*num_features), num_movies, num_features);
Theta = reshape(theta(num_movies*num_features+1:end), num_users, num_features);

fprintf('Recommender system learning completed.\n');
fprintf('\nProgram paused. Press enter to continue.\n');
pause;

p = X * Theta';
my_predictions = p(:,1) + Ymean;
movieList = loadMovieList();
[r, ix] = sort(my_predictions, 'descend');
fprintf('\nTop recommendations for you:\n');
for i=1:10
  j = ix(i);
  fprintf('Predicting rating %.1f for movie %s\n', my_predictions(j), movieList{j});
endfor

fprintf('\n\nOriginal ratings provided:\n');
for i = 1:length(my_ratings)
  if my_ratings(i) > 0 
    fprintf('Rated %d for %s\n', my_ratings(i), movieList{i});
  endif
endfor

File: 11-Recommender System/ex11.py
import numpy as np
from scipy import io, optimize
import matplotlib.pyplot as plt

def plot_data(Y):
    plt.imshow(Y, aspect='auto')
    plt.ylabel('Movies')
    plt.xlabel('Users')

def collaborative_cost_function(params, Y, R, num_users, num_movies, num_features, lamda):
    X = params[0:num_movies * num_features].reshape((num_movies, num_features))
    theta = params[num_movies * num_features:].reshape((num_users, num_features))

    J = 0.5 * np.sum(np.sum(R * np.square(X.dot(theta.T) - Y)))
    X_grad = (R * (X.dot(theta.T) - Y)).dot(theta)
    theta_grad = (R * (X.dot(theta.T) - Y)).T.dot(X)

    J = J + 0.5*lamda*np.sum(np.square(theta)) + 0.5*lamda*np.sum(np.square(X))
    X_grad = X_grad + lamda*X
    theta_grad = theta_grad + lamda*theta
    grad = np.hstack((X_grad.ravel(), theta_grad.ravel()))
    return J, grad

def load_movie_list():
    movie_list = []
    with open("movie_ids.txt") as f:
        for line in f:
            movie_list.append(line[line.index(' ') + 1:].rstrip())
    return movie_list

def normalize_ratings(Y, R):
    m = Y.shape[0]
    Y_mean = np.zeros(m)
    Y_norm = np.zeros(Y.shape)
    for i in range(m):
        idx = np.nonzero(R[i,] == 1)
        Y_mean[i] = np.mean(Y[i, idx])
        Y_norm[i, idx] = Y[i, idx] - Y_mean[i]
    return Y_norm, Y_mean

if __name__ == "__main__":

    # =============== Part 1: Loading movie ratings dataset ================
    print ('Loading movie ratings dataset.')
    data = io.loadmat('ex8_movies.mat')

    # Y is a 1682x943 matrix, containing ratings (1-5) of 1682 movies on 943 users
    Y = data['Y']
    # R is a 1682x943 matrix, where R(i,j) = 1 if and only if user j gave a rating to movie i
    R = data['R']

    print ('Average rating for movie 1 (Toy Story): {}/5'.format(np.mean(Y[0, np.nonzero(R[0,])])))
    plt.figure(1)
    plot_data(Y)

    # ============ Part 2: Collaborative Filtering Cost Function ===========
    # Load pre-trained weights (X, Theta, num_users, num_movies, num_features)
    data = io.loadmat('ex8_movieParams.mat')
    X = data['X']
    theta = data['Theta']
    num_users = data['num_users'].ravel()[0]
    num_movies = data['num_movies'].ravel()[0]
    num_features = data['num_features'].ravel()[0]

    # Reduce the data set size so that this runs faster
    num_users = 4
    num_movies = 5
    num_features = 3
    X = X[0:num_movies, 0:num_features]
    theta = theta[0:num_users, 0:num_features]
    Y = Y[0:num_movies, 0:num_users]
    R = R[0:num_movies, 0:num_users]

    J, grad = collaborative_cost_function(np.hstack((X.flatten(), theta.flatten())),
                                          Y, R, num_users, num_movies, num_features, lamda=1.5)
    print ('Cost at loaded parameters:', J)

    # ============== Part 3: Entering ratings for a new user ===============
    movie_list = load_movie_list()

    # Initialize my ratings
    my_ratings = np.zeros(len(movie_list), dtype=np.int)
    my_ratings[0] = 4
    my_ratings[97] = 2
    my_ratings[6] = 3
    my_ratings[11] = 5
    my_ratings[53] = 4
    my_ratings[63] = 5
    my_ratings[65] = 3
    my_ratings[68] = 5
    my_ratings[182] = 4
    my_ratings[225] = 5
    my_ratings[354] = 5

    print ('New user ratings:')
    for i in np.argwhere(my_ratings > 0).ravel():
        print ('Rated {} for {}'.format(my_ratings[i], movie_list[i]))

    # ================== Part 4: Learning Movie Ratings ====================
    print ('Training collaborative filtering...')
    data = io.loadmat('ex8_movies.mat')

    # Y is a 1682x943 matrix, containing ratings (1-5) of 1682 movies on 943 users
    Y = data['Y']
    # R is a 1682x943 matrix, where R(i,j) = 1 if and only if user j gave a rating to movie i
    R = data['R']

    # Add our own ratings to the data matrix
    Y = np.hstack((my_ratings.reshape(len(movie_list), 1), Y))
    R = np.hstack((my_ratings.reshape(len(movie_list), 1) != 0, R))

    # Normalize Ratings
    Y_norm, Y_mean = normalize_ratings(Y, R)

    num_users = Y.shape[1]
    num_movies = Y.shape[0]
    num_features = 10

    # Set Initial Parameters (Theta, X)
    X = np.random.randn(num_movies, num_features)
    theta = np.random.randn(num_users, num_features)

    initial_parameters = np.hstack((X.flatten(), theta.flatten()))

    # Set Regularization
    l = 10
    result = optimize.minimize(fun=collaborative_cost_function, x0=initial_parameters,
                                args=(Y_norm, R, num_users, num_movies, num_features, l),
                                method='CG', jac=True, options={'maxiter': 150})

    X = result.x[0:num_movies * num_features].reshape((num_movies, num_features))
    theta = result.x[num_movies * num_features:].reshape((num_users, num_features))

    print ('Recommender system learning completed.')

    # ================== Part 5: Recommendation for you ====================
    p = X.dot(theta.T)
    my_predictions = p[:, 0] + Y_mean
    idx = np.argsort(my_predictions)[::-1]
    print ('Top recommendations for you:')
    for i in range(10):
        print ('Predicting rating {0:.1f} for movie {1:s}'.format(my_predictions[idx[i]], movie_list[idx[i]]))

    print ('Original ratings provided:')
    for i in np.argwhere(my_ratings > 0).ravel():
        print ('Rated {} for {}'.format(my_ratings[i], movie_list[i]))
